hash/hash_comparator.go

```diff
@@ -558,3 +576,78 @@ func decodeMD5Hash(encodedHash string) (pf, salt, hash []byte, err error) {
 		return nil, nil, nil, ErrInvalidHash
 	}
 }
+
+// decodeLegacyHash decodes argon2id wrapped legacy password hash.
+// format: $legacy$alg=<md5|sha1|sha256|sha512>,format=<base64 encoded format, e.g: '{PEPPER}-{PWD}-{SALT}'>,i=<number of iterations>,(salt=<base64 encoded salt>,pepper=<base64 encoded pepper>)$<argon2id encoded hash>
+func decodeLegacyHash(encodedHash string) (phc, alg, format string, iterations int, salt, pepper []byte, err error) {
+	parts := strings.SplitN(encodedHash, "$", 4)
+	if len(parts) != 4 {
+		return "", "", "", 0, nil, nil, ErrInvalidHash
+	}
+	phc = "$" + parts[3]
+	legacyParams := strings.Split(parts[2], ",")
+	for _, lp := range legacyParams {
+		switch kv := strings.Split(lp, "="); kv[0] {
+		case "alg":
+			alg = kv[1]
+		case "format":
+			f, err := base64.RawStdEncoding.DecodeString(kv[1])
+			if err != nil {
+				return "", "", "", 0, nil, nil, ErrInvalidHash
+			}
+			format = string(f)
+		case "i":
+			iterations, err = strconv.Atoi(kv[1])
+			if err != nil {
+				return "", "", "", 0, nil, nil, ErrInvalidHash
+			}
+		case "salt":
+			salt, err = base64.RawStdEncoding.DecodeString(kv[1])
+			if err != nil {
+				return "", "", "", 0, nil, nil, ErrInvalidHash
+			}
+		case "pepper":
+			pepper, err = base64.RawStdEncoding.DecodeString(kv[1])
+			if err != nil {
+				return "", "", "", 0, nil, nil, ErrInvalidHash
+			}
+		}
+	}
+	return
+}
+
+func legacyHash(alg, format string, iterations int, salt, pepper, pwd []byte) ([]byte, error) {
+	if len(pwd) < 1 || !strings.Contains(format, "{PWD}") || iterations < 1 || !(alg == "md5" || alg == "sha1" || alg == "sha256" || alg == "sha512") {
+		return nil, ErrInvalidHash
+	}
+	replacements := []string{"{PWD}", string(pwd)}
+	if salt != nil {
+		replacements = append(replacements, "{SALT}", string(salt))
+	}
+	if pepper != nil {
+		replacements = append(replacements, "{PEPPER}", string(pepper))
+	}
+	toHash := []byte(strings.NewReplacer(replacements...).Replace(format))
+
+	var hash []byte
+	for i := 1; i <= iterations; i++ {
+		switch alg {
+		case "md5":
+			sum := md5.Sum(toHash) //#nosec G401 -- compatibility for imported passwords
+			hash = sum[:]
+		case "sha1":
+			sum := sha1.Sum(toHash) //#nosec G401 -- compatibility for imported passwords
+			hash = sum[:]
+		case "sha256":
+			sum := sha256.Sum256(toHash)
+			hash = sum[:]
+		case "sha512":
+			sum := sha512.Sum512(toHash)
+			hash = sum[:]
+		default:
+			return nil, ErrUnknownHashAlgorithm
+		}
+		toHash = hash
+	}
+	return hash, nil
```

Comment:
I think this is a bit easier to understand:

```suggestion
	toHash := []byte(strings.NewReplacer(replacements...).Replace(format))

	var hasher hash.Hash
	switch alg {
	case "md5":
		hasher = md5.New() //#nosec G401 -- compatibility for imported passwords
	case "sha1":
		hasher = sha1.New() //#nosec G401 -- compatibility for imported passwords
	case "sha256":
		hasher = sha256.New()
	case "sha512":
		hasher = sha512.New()
	default:
		return nil, ErrUnknownHashAlgorithm
	}
	for i := 1; i <= iterations; i++ {
		toHash = hasher.Sum(toHash)
	}
	return toHash, nil
```

Reply:
`toHash = hasher.Sum(toHash)` does not do what you and I think 😄  Should be:
```
hasher.Write(toHash)
toHash = hasher.Sum([]byte{})
hasher.Reset()
```